tests/frontend/test_utils.py

```diff
@@ -1,686 +1,660 @@
+        assert prog.circuit[0].op.p[0].x == rval
+
+        # check second queue op is BSgate(thetaval, phival)
+        assert isinstance(prog.circuit[1].op, BSgate)
+        assert prog.circuit[1].reg == list(q)
+        assert prog.circuit[1].op.p[0].x == thetaval
+        assert prog.circuit[1].op.p[1].x == phival
+
+    def test_multimode_wrong_num_modes_apply_operation(self, prog):
+        """Test exceptions raised when applying an operation to
+        multiple modes with incorrect num modes"""
+        rval = 4.32
+        thetaval = -0.654
+        phival = 0.543
+
+        @utils.operation(1)
+        def dummy_func(r, theta, phi, q):
+            Sgate(r) | q[0]
+            BSgate(theta, phi) | (q[0], q[1])
+
+        with prog.context as q:
+            with pytest.raises(ValueError, match="Wrong number of subsystems"):
+                dummy_func(rval, thetaval, phival) | q
+
+    def test_single_mode_no_args_apply_operation(self, prog):
+        """Test applying an operation to a single mode"""
+        rval = 4.32
+
+        @utils.operation(2)
+        def dummy_func(q):
+            Sgate(rval) | q[0]
+
+        with prog.context as q:
+            dummy_func() | q
+
+        # check eng queue matches the operation
+        assert len(prog) == 1
+
+        # check first queue op is Sgate(rval)
+        assert isinstance(prog.circuit[0].op, Sgate)
+        assert prog.circuit[0].reg == [q[0]]
+        assert prog.circuit[0].op.p[0].x == rval
+
+    def test_multimode_args_apply_operation(self, prog):
+        """Test applying an operation to multiple modes"""
+        rval = 4.32
+        thetaval = -0.654
+        phival = 0.543
+
+        @utils.operation(2)
+        def dummy_func(r, theta, phi, q):
+            Sgate(r) | q[0]
+            BSgate(theta, phi) | (q[0], q[1])
+
+        with prog.context as q:
+            with pytest.raises(ValueError, match="Wrong number of subsystems"):
+                dummy_func(rval, thetaval, phival) | q[0]
+
+        with prog.context as q:
+            dummy_func(rval, thetaval, phival) | q
+
+        # check eng queue matches the operation
+        assert len(prog) == 2
+
+        # check first queue op is Sgate(rval)
+        assert isinstance(prog.circuit[0].op, Sgate)
+        assert prog.circuit[0].reg == [q[0]]
+        assert prog.circuit[0].op.p[0].x == rval
+
+        # check second queue op is BSgate(thetaval, phival)
+        assert isinstance(prog.circuit[1].op, BSgate)
+        assert prog.circuit[1].reg == list(q)
+        assert prog.circuit[1].op.p[0].x == thetaval
+        assert prog.circuit[1].op.p[1].x == phival
+
+
+# ===================================================================================
+# Engine utility tests
+# ===================================================================================
+
+
+class TestEngineUtilityFunctions:
+    """Tests for some engine auxiliary functions"""
+
+    @pytest.mark.skip('FIXME is_unitary')
```

Comment:
Can we make a separate ticket/issue for specific related fixes that need to be made going forward (e.g., I've seen `is_unitary`, `is_channel`, and one about hbar issues)

Reply:
Sure